### PR TITLE
Add Node Selector via config

### PR DESCRIPTION
**Pull request by andrew-waters**
Signed-off-by: Andrew Waters <andrew-waters@users.noreply.github.com>

#### What this PR does

Adds the ability to move microservices to a dedicated node selector

#### Which issue(s) this PR fixes or relates to

No issue, found and fixed

#### Checklist

- [x] ~Tests updated~
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by osg-grafana** [comment]
Besides the CHANGELOG, does an operator need more documentation to know that this exist over time and be able to use it?

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 ## Grafana Mimir - main / unreleased
 
+* [ENHANCEMENT] Microservices: Ability to add NodeSelector
```
Comment (56quarters) [inline CHANGELOG.md:5]: Can you move this to the `## Jsonnet` section?

**Review by 56quarters — APPROVED** [review]
LGTM besides one small change but you'll have to rebase on latest `main` due to some recent CI changes

**Comment by pstibrany** [comment]
Could you please rebase your PR on top of latest `main`? That should fix the problem with CI checks failing. Thank you!

File: operations/mimir/alertmanager.libsonnet [hunk 2]
```diff
@@ -75,6 +75,7 @@
       statefulSet.mixin.spec.updateStrategy.withType('RollingUpdate') +
       statefulSet.mixin.spec.template.spec.withTerminationGracePeriodSeconds(900) +
       $.util.configVolumeMount($._config.overrides_configmap, $._config.overrides_configmap_mountpoint) +
+      (if ! std.isObject($._config.node_selector) then {} else statefulSet.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```
Comment (56quarters) [inline operations/mimir/alertmanager.libsonnet:78]:
```suggestion
      (if !std.isObject($._config.node_selector) then {} else statefulSet.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```

File: operations/mimir/compactor.libsonnet [hunk 3]
```diff
@@ -115,6 +115,7 @@
     statefulSet.mixin.spec.template.spec.securityContext.withRunAsUser(0) +
     statefulSet.mixin.spec.updateStrategy.withType('RollingUpdate') +
     statefulSet.mixin.spec.template.spec.withTerminationGracePeriodSeconds(900) +
+    (if ! std.isObject($._config.node_selector) then {} else statefulSet.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```
Comment (56quarters) [inline operations/mimir/compactor.libsonnet:118]:
```suggestion
    (if !std.isObject($._config.node_selector) then {} else statefulSet.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```

File: operations/mimir/distributor.libsonnet [hunk 4]
```diff
@@ -54,6 +54,7 @@
     deployment.new('distributor', 3, [$.distributor_container]) +
     (if $._config.distributor_allow_multiple_replicas_on_same_node then {} else $.util.antiAffinity) +
     $.util.configVolumeMount($._config.overrides_configmap, $._config.overrides_configmap_mountpoint) +
+    (if ! std.isObject($._config.node_selector) then {} else deployment.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```
Comment (56quarters) [inline operations/mimir/distributor.libsonnet:57]:
```suggestion
    (if !std.isObject($._config.node_selector) then {} else deployment.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```

File: operations/mimir/flusher-job-blocks.libsonnet [hunk 5]
```diff
@@ -44,6 +44,7 @@
     job.mixin.spec.template.metadata.withLabels({ name: 'flusher' }) +
     job.mixin.spec.template.spec.securityContext.withRunAsUser(0) +
     job.mixin.spec.template.spec.withTerminationGracePeriodSeconds(300) +
+    (if ! std.isObject($._config.node_selector) then {} else job.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```
Comment (56quarters) [inline operations/mimir/flusher-job-blocks.libsonnet:47]:
```suggestion
    (if !std.isObject($._config.node_selector) then {} else job.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```

File: operations/mimir/ingester.libsonnet [hunk 6]
```diff
@@ -93,6 +93,7 @@
     // rolled out one by one (the next pod will be rolled out once the previous is
     // ready).
     statefulSet.mixin.spec.withPodManagementPolicy('Parallel') +
+    (if ! std.isObject($._config.node_selector) then {} else statefulSet.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```
Comment (56quarters) [inline operations/mimir/ingester.libsonnet:96]:
```suggestion
    (if !std.isObject($._config.node_selector) then {} else statefulSet.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```

File: operations/mimir/memcached.libsonnet [hunk 7]
```diff
@@ -14,6 +14,7 @@ memcached {
         self.memcached_exporter,
       ], []) +
       statefulSet.mixin.spec.withServiceName(self.name) +
+      (if ! std.isObject($._config.node_selector) then {} else statefulSet.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```
Comment (56quarters) [inline operations/mimir/memcached.libsonnet:17]:
```suggestion
      (if !std.isObject($._config.node_selector) then {} else statefulSet.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```

File: operations/mimir/multi-zone.libsonnet [hunk 8]
```diff
@@ -94,6 +94,7 @@
     statefulSet.mixin.spec.updateStrategy.withType('OnDelete') +
     statefulSet.mixin.spec.template.spec.withTerminationGracePeriodSeconds(1200) +
     statefulSet.mixin.spec.withReplicas(std.ceil($._config.multi_zone_ingester_replicas / 3)) +
+    (if ! std.isObject($._config.node_selector) then {} else statefulSet.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```
Comment (56quarters) [inline operations/mimir/multi-zone.libsonnet:97]:
```suggestion
    (if !std.isObject($._config.node_selector) then {} else statefulSet.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```

File: operations/mimir/querier.libsonnet [hunk 9]
```diff
@@ -53,6 +53,7 @@
     deployment.new(name, $._config.querier.replicas, [container]) +
     (if $._config.querier_allow_multiple_replicas_on_same_node then {} else $.util.antiAffinity) +
     $.util.configVolumeMount($._config.overrides_configmap, $._config.overrides_configmap_mountpoint) +
+    (if ! std.isObject($._config.node_selector) then {} else deployment.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```
Comment (56quarters) [inline operations/mimir/querier.libsonnet:56]:
```suggestion
    (if !std.isObject($._config.node_selector) then {} else deployment.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```

File: operations/mimir/query-frontend.libsonnet [hunk 10]
```diff
@@ -45,6 +45,7 @@
     deployment.new(name, $._config.queryFrontend.replicas, [container]) +
     $.util.configVolumeMount($._config.overrides_configmap, $._config.overrides_configmap_mountpoint) +
     (if $._config.query_frontend_allow_multiple_replicas_on_same_node then {} else $.util.antiAffinity) +
+    (if ! std.isObject($._config.node_selector) then {} else deployment.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```
Comment (56quarters) [inline operations/mimir/query-frontend.libsonnet:48]:
```suggestion
    (if !std.isObject($._config.node_selector) then {} else deployment.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```

File: operations/mimir/query-scheduler.libsonnet [hunk 11]
```diff
@@ -25,6 +25,7 @@
   newQuerySchedulerDeployment(name, container)::
     deployment.new(name, 2, [container]) +
     $.util.configVolumeMount($._config.overrides_configmap, $._config.overrides_configmap_mountpoint) +
+    (if ! std.isObject($._config.node_selector) then {} else deployment.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```
Comment (56quarters) [inline operations/mimir/query-scheduler.libsonnet:28]:
```suggestion
    (if !std.isObject($._config.node_selector) then {} else deployment.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```

File: operations/mimir/query-tee.libsonnet [hunk 12]
```diff
@@ -21,6 +21,7 @@
     $.jaeger_mixin,
 
   query_tee_deployment: if !($._config.query_tee_enabled) then {} else
+    (if ! std.isObject($._config.node_selector) then {} else deployment.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```
Comment (56quarters) [inline operations/mimir/query-tee.libsonnet:24]:
```suggestion
    (if !std.isObject($._config.node_selector) then {} else deployment.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```

File: operations/mimir/ruler.libsonnet [hunk 13]
```diff
@@ -53,6 +53,7 @@
   ruler_deployment:
     if $._config.ruler_enabled then
       deployment.new('ruler', 2, [$.ruler_container]) +
+      (if ! std.isObject($._config.node_selector) then {} else deployment.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```
Comment (56quarters) [inline operations/mimir/ruler.libsonnet:56]:
```suggestion
      (if !std.isObject($._config.node_selector) then {} else deployment.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```

File: operations/mimir/store-gateway.libsonnet [hunk 14]
```diff
@@ -82,6 +82,7 @@
     statefulSet.mixin.spec.template.metadata.withLabels({ name: name }) +
     statefulSet.mixin.spec.selector.withMatchLabels({ name: name }) +
     statefulSet.mixin.spec.template.spec.securityContext.withRunAsUser(0) +
+    (if ! std.isObject($._config.node_selector) then {} else statefulSet.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```
Comment (56quarters) [inline operations/mimir/store-gateway.libsonnet:85]:
```suggestion
    (if !std.isObject($._config.node_selector) then {} else statefulSet.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
```

**Comment by andrew-waters** [comment]
Thanks for the suggestions on that listing @56quarters, all committed now. Hopefully we're now 🟢

**Comment by 56quarters** [comment]
Fixed up the last few lint errors. Thanks again!

**Pull request merged**